src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,358 @@
+
+            return _options.CpuUsageWeight * normalizedCpuUsage +
+                   _options.MemoryUsageWeight * normalizedMemoryUsage +
+                   _options.AvailableMemoryWeight * normalizedAvailableMemory +
+                   _options.AvailableMemoryWeight * normalizedPhysicalMemory;
+        }
+
+        return _options.CpuUsageWeight * normalizedCpuUsage;
+    }
+
+    public void RemoveSilo(SiloAddress address)
+         => _siloStatistics.TryRemove(address, out _);
+
+    public void SiloStatisticsChangeNotification(SiloAddress address, SiloRuntimeStatistics statistics)
+        => _siloStatistics.AddOrUpdate(
+            address,
+            addValueFactory: static (_, statistics) => new (statistics),
+            updateValueFactory: static (_, existing, statistics) =>
+            {
+                existing.Update(statistics);
+                return existing;
+            },
+            statistics);
+
+    private readonly record struct ResourceStatistics(float? CpuUsage, float? AvailableMemory, long? MemoryUsage, long? TotalPhysicalMemory, bool IsOverloaded);
+
+    private sealed class FilteredSiloStatistics(SiloRuntimeStatistics statistics)
+    {
+        private readonly DualModeKalmanFilter<float> _cpuUsageFilter = new();
+        private readonly DualModeKalmanFilter<float> _availableMemoryFilter = new();
+        private readonly DualModeKalmanFilter<long> _memoryUsageFilter = new();
+
+        private float? _cpuUsage = statistics.CpuUsage;
+        private float? _availableMemory = statistics.AvailableMemory;
+        private long? _memoryUsage = statistics.MemoryUsage;
+        private long? _totalPhysicalMemory = statistics.TotalPhysicalMemory;
+        private bool _isOverloaded = statistics.IsOverloaded;
+
+        public ResourceStatistics Value => new(_cpuUsage, _availableMemory, _memoryUsage, _totalPhysicalMemory, _isOverloaded);
+
+        public void Update(SiloRuntimeStatistics statistics)
+        {
+            _cpuUsage = _cpuUsageFilter.Filter(statistics.CpuUsage);
+            _availableMemory = _availableMemoryFilter.Filter(statistics.AvailableMemory);
+            _memoryUsage = _memoryUsageFilter.Filter(statistics.MemoryUsage);
+            _totalPhysicalMemory = statistics.TotalPhysicalMemory;
+            _isOverloaded = statistics.IsOverloaded;
+        }
+    }
+
+    private sealed class DualModeKalmanFilter<T> where T : unmanaged, INumber<T>
+    {
+        private KalmanFilter _slowFilter = new();
+        private KalmanFilter _fastFilter = new();
+        
+        private FilterRegime _regime = FilterRegime.Slow;
+
+        private enum FilterRegime
+        {
+            Slow,
+            Fast
+        }
+
+        public T Filter(T? measurement)
+        {
+            T _measurement = measurement ?? T.Zero;
+
+            T slowCovariance = T.Zero;
+            T fastCovariance = T.CreateChecked(0.01);
```

Comment:
For the case where `T` is `long`, this will be zero. Is that desirable? Perhaps we should convert to a floating point number instead?

Reply:
No its not desirable, I missed that as I was testing only with `float`'s